### PR TITLE
Finalize methods that do not support overrides

### PR DESCRIPTION
**Pull request by dduugg**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
I noticed some methods with comments discouraging overrides. Let's enforce that in code instead.

See: https://sorbet.org/docs/final

### CODE REVIEW

File: Library/Homebrew/requirement.rb [hunk 1]
```diff
@@ -92,8 +99,15 @@ def satisfied_result_parent
...
-  # Overriding {#modify_build_environment} is unsupported.
-  # Pass a block to the env DSL method instead.
+  # Pass a block to the env DSL method instead of overriding.
+  sig(:final) {
+    params(
+      env:          T.nilable(String),
+      cc:           T.nilable(String),
+      build_bottle: T::Boolean,
+      bottle_arch:  T.nilable(String),
+    ).returns(T.untyped)
```
Comment (dduugg) [inline Library/Homebrew/requirement.rb:109]: Note that this method currently returns `nil` or `T::Private::Types::Void::VOID`, which is clumsy/incorrect to encode in a type signature. However, we currently have a [test](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/requirement_spec.rb#L178) that relies on this being nil/non-nil depending on circumstance, so I didn't undertake a refactor. Feel free to weigh in if you want this to resolved by other means though.

File: Library/Homebrew/requirement.rb [hunk 2]
```diff
@@ -92,8 +99,15 @@ def satisfied_result_parent
...
 
-  # Overriding {#modify_build_environment} is unsupported.
-  # Pass a block to the env DSL method instead.
+  # Pass a block to the env DSL method instead of overriding.
+  sig(:final) {
+    params(
+      env:          T.nilable(String),
+      cc:           T.nilable(String),
+      build_bottle: T::Boolean,
+      bottle_arch:  T.nilable(String),
```
Comment (dduugg) [inline Library/Homebrew/requirement.rb:108]: params are based on https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/ENV.rb#L36-L39 (though I noticed that `build_bottle` is `T.nilable(T::Boolean)` in some places)

Comment (reitermarkus) [reply Library/Homebrew/requirement.rb:108]: I think `T::Boolean` is preferable in most cases. We should adjust the call sites to avoid `T.nilable(T::Boolean)` being passed where this is still the case.

File: Library/Homebrew/requirement.rb (hunk 1 above)
Comment (reitermarkus) [reply Library/Homebrew/requirement.rb:109]: I think this is kinda weird, since `ENV#prepend_path` can technically return `nil` masked as `T::Private::Types::Void::VOID`, right?

Comment (reitermarkus) [reply Library/Homebrew/requirement.rb:109]: I'd say this should always return `nil`, or we just remove the test and make it `.void`.

Probably the latter since we don't have any test/usage that uses the return value anyways.

Comment (MikeMcQuaid) [reply Library/Homebrew/requirement.rb:109]: Agreed that we should probably update the test here.

**Comment by MikeMcQuaid** [comment]
@dduugg what's the latest here?

File: Library/Homebrew/requirement.rb (hunk 1 above)
Comment (dduugg) [reply Library/Homebrew/requirement.rb]: I made the return type consistently void, and also updated the test to assert no error occurs (very roughly equivalent to checking for `nil` return, in this case).

**Comment by dduugg** [comment]
> @dduugg what's the latest here?

Sorry for the latency here, PTAL. I've updated the signature and test per comments.

**Review by MikeMcQuaid — APPROVED** [review]
Looks good, thanks for the quick update @dduugg!
